core/src/types/blocking_read/buffer_iterator.rs

```diff
@@ -0,0 +1,115 @@
+use std::ops::Range;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+use bytes::Buf;
+use rayon::prelude::*;
+
+use crate::raw::*;
+use crate::*;
+
+pub(super) struct BufferIterator {
+    inner: Arc<dyn oio::BlockingRead>,
+    it: RangeIterator,
+
+    chunk: usize,
+    end: Option<u64>,
+    concurrent: usize,
+    finished: Arc<AtomicBool>,
+}
+
+impl BufferIterator {
+    pub fn new(
+        inner: Arc<dyn oio::BlockingRead>,
+        chunk: Option<usize>,
+        concurrent: usize,
+        offset: u64,
+        end: Option<u64>,
+    ) -> Self {
+        let chunk = chunk.unwrap_or(4 * 1024 * 1024);
+        let it = RangeIterator {
+            offset,
+            chunk: chunk as u64,
+        };
+        Self {
+            inner,
+            it,
+            chunk,
+            end,
+            concurrent,
+            finished: Arc::new(AtomicBool::new(false)),
+        }
+    }
+}
+
+impl Iterator for BufferIterator {
+    type Item = Result<Buffer>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.finished.load(Ordering::Relaxed) {
+            return None;
+        }
+
+        let mut bufs = Vec::with_capacity(self.concurrent);
+
+        let intervals: Vec<Range<u64>> = (0..self.concurrent)
```

Comment:
Maybe we should put `self.finished` inside the `RangeIterator` should it can know when to stop returning new ranges.

Reply:
> Maybe we should put `self.finished` inside the `RangeIterator` should it can know when to stop returning new ranges.

When `end` is None, We only know end when after read.

Reply:
> When `end` is None, We only know end when after read.

That's why we need to store `finished` in `RangeIterator`.

Reply:
Do you mean `read` also put into RangeIterator?

Reply:
@Xuanwo I submit a new one, is this what you want?

Reply:
Close but not what I want. Can I push changes to this PR directly?

Reply:
> Close but not what I want. Can I push changes to this PR directly?

Of course, thank for your patient.

Reply:
@Xuanwo Do you remember this?